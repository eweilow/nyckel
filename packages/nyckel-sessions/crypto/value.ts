import { saltKey } from "./salt";
import { encrypt, decrypt } from "./aes256";

function password(id: string, salt: string, secret: string) {
  return saltKey(id, salt, secret);
}

export function encryptData<T>(
  id: string,
  salt: string,
  secret: string,
  data: T
): string {
  const key = password(id, salt, secret);
  return encrypt(key, JSON.stringify({ data }));
}

export function decryptData<T>(
  id: string,
  salt: string,
  secret: string,
  encrypted: string
): T | null {
  try {
    const key = password(id, salt, secret);
    const decrypted = decrypt(key, encrypted);

    const { data } = JSON.parse(decrypted) as {
      data: T;
    };

    return data;
  } catch (err) {
    console.error(err);
    return null;
  }
}
