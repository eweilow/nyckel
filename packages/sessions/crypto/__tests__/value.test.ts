import { decryptData, encryptData } from "../value";

describe("encryption of whole data", () => {
  it("encrypts and decrypts correctly", () => {
    const id = "id";
    const salt = "salt";
    const secret = "secret";

    const original = {
      a: 1,
      b: 2,
      c: {
        d: 3
      }
    };

    const encrypted = encryptData(id, salt, secret, original);
    const decrypted = decryptData(id, salt, secret, encrypted);

    expect(original).toEqual(decrypted);
  });

  it("returns null and logs error in console if data is invalid", () => {
    const consoleSpy = jest
      .spyOn(console, "error")
      .mockImplementation(() => {});

    const id = "id";
    const salt = "salt";
    const secret = "secret";

    const original = {
      a: 1,
      b: 2,
      c: {
        d: 3
      }
    };

    const encrypted = encryptData(id, salt, secret, original);
    const decrypted = decryptData(id, salt, secret, "s" + encrypted);

    expect(original).not.toEqual(decrypted);
    expect(decrypted).toEqual(null);
    expect(consoleSpy.mock.calls).toMatchSnapshot();
    consoleSpy.mockClear();
  });
});
