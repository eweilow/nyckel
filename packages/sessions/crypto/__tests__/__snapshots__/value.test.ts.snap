// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encryption of whole data returns null and logs error in console if data is invalid 1`] = `
Array [
  Array [
    [SyntaxError: Unexpected token I in JSON at position 0],
  ],
]
`;
